### ISSUE TITLE
Bandlimited pulse oscillator with a-rate pulseWidth parameter

### ISSUE THREAD

**Issue by svgeesus** [WebAudio/web-audio-api-v2#7]
Unlike the existing square wave, here the mark-space ration is changeable and modulatable. This is a classic effect and (unlike a static mark:space ratio) cannot be achieved using the FFT arrays on a custom oscillator node.

This was asked for several times at WAC2016

**Comment by svgeesus** [comment]
How much detail should the specification have about the implementation? Obviously it needs to say the output must be bandlimited but do we want to specify BLIT or leave it up to implementations? OscillatorNode is fairly specific on details.

**Comment by rtoy** [comment]
Teleconf:  Would be really nice to have a polyfill with a Worklet to see and hear what the oscillator does.

**Comment by charlieroberts** [comment]
just a quick note that FM PWM is a nice option... I use polyblep for other waveforms in gibber, but FM for PWM because it just seems to sound better.to me subjectively. http://scp.web.elte.hu/papers/synthesis1.pdf

Although I noticed that @svgeesus linked a paper on oscillators with hard sync which I don't think would work using an FM approach (I could be wrong about this). Would love to see hard sync support in oscillator nodes some day...

**Comment by skratchdot** [comment]
I created a demo here:
https://github.com/skratchdot/web-audio-api-v2-issue-7

But it's not quite right. I need to read some of the articles posted, and read more about the math / audio dsp side of things.  I'm coming from a "web dev" perspective, so am probably missing something simple.

Anyways, if anyone wants to fix this "overly complex" demo, I think you should just be able to edit the worklet code here:
https://github.com/skratchdot/web-audio-api-v2-issue-7/blob/master/public/pulse-oscillator.js

Also, from my understanding, we should be able to create a custom oscillator that is a pulse oscillator with a custom pulseWidth- but if we do that, we can't automate the pulseWidth parameter, correct?  (because there is no way to automate a "setPeriodicWave", etc).

Also, when I try to create a "custom" OscillatorNode with a variable pulseWidth, the images change drastically depending on the `disableNormalization` setting.

# disableNormalization=false
![image](https://user-images.githubusercontent.com/434470/90324716-3cd95b00-df40-11ea-8a34-e5b5d7588ab2.png)

# disableNormalization=true
![image](https://user-images.githubusercontent.com/434470/90324722-4367d280-df40-11ea-88d1-2c349aa2fecf.png)

**Comment by rtoy** [comment]
From today's teleconf, a simple design was done:
```
interface PulseWidthOscillatorNode : AudioScheduledSourceNode {
  constructor (BaseAudioContext context, optional PulseWidthOscillatorOptions options = {});
  readonly attribute AudioParam frequency;
  readonly attribute AudioParam detune;
  readonly attribute AudioParam pulseWidth;
};

dictionary PulseWidthOscillatorOptions : AudioNodeOptions {
  float frequency = 440;
  float detune = 0;
  float pulseWidth = 0.5;
};
```

This oscillator is bandlimited and goes between -1 and 1.  The pulseWidth range is `[0,1]`, where a value of 0 means the output is a constant -1 and a value of 1 means the output is a constant +1.  A value of 0.5 is a square wave.

**Comment by Rumyra** [comment]
Would we need pulse offset here as per https://github.com/WebAudio/web-audio-api-v2/issues/1?

Came up in wg/cg meeting

**Comment by rtoy** [comment]
Teleconf: Also what about hard sync?  We should be consistent with Oscillator

**Comment by rtoy** [comment]
Oh, I think I got the output inverted.  To match a square wave, the output should start at +1 and go to -1.  So a pulseWidth near 0 is a little bit of +1 followed by mostly -1; a width close to 1 is mostly +1 followed by a little bit of -1.

**Comment by rtoy** [comment]
Teleconf:  Need to resolve #1 and #9 before we can finalize this.
